### PR TITLE
DeSo V3 Messages

### PR DESCRIPTION
**Pull request by AeonSw4n**

### CODE REVIEW

File: dips/dip-5.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+- [DeSo core](https://github.com/deso-protocol/core/pull/170)
+- [DeSo backend](https://github.com/deso-protocol/backend/pull/253)
+- [DeSo identity](https://github.com/deso-protocol/identity/pull/107)
+- [DeSo node frontend](https://github.com/deso-protocol/frontend/pull/543)
+
+### Technical changes
+
+#### Core
+A conditional statement will be added in `_connectBasicTransfer()` to look for transaction's `ExtraData` and check for three fields: `MessagingPublicKey`, `MessagingKeyName`, and `MessagingKeySignature`. These respectively will contain the public key and its name along with an owner-signed hash of `(MessagingPublicKey + MessagingKeyName)`. The signature is provided because we want derived key holders to submit messaging keys; therefore, the transaction signature alone wouldn't be sufficient to prove authorization of the messaging key. Similarly, disconnecting a messaging key will happen in `_disconnectBasicTransfer()` and will delete the messaging entry from the `UtxoView` and DB. Core will only accept one key for a label, an attempt to submit a different public key for the same key name will silently error.
```
Comment (maebeam) [inline dips/dip-5.md:36]: I don't like the precedent this sets. Avoiding a fork by shoving information into a basic transfer's ExtraData map does not seem advisable. I would much prefer we just add a new transaction type and schedule a network upgrade.

Comment (maebeam) [reply dips/dip-5.md:36]: Also, if nodes are going to build these indexes correctly, we need them to upgrade accordingly so we may as well set a block height and schedule this change for the next fork.

File: dips/dip-5.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
...ill be using the V2 message protocol with shared secrets between their primary keys. But as soon as a user registers a messaging key named "default-key", the DeSo Identity Service will be encrypting messages to this public key. The "default-key" messaging key will be returned in `/derive` Window API endpoint. It is suggested to pass the messaging key information in the `ExtraData` of the transaction authorizing a derived key. We also modify the `encrypt` and `decrypt` methods in the iframe API. The `encrypt` method will make a request to the Backend API to look for sender's and recipient's default messaging keys. If either messaging key exists, `encrypt` will create a V3 message using one or both of the messaging keys and return messaging party information along with the encrypted message. The `decrypt` method will make no additional Backend API requests, and will require the messaging party to be passed along with the encrypted message. We make a modification to the request payload to `decrypt` to require `Version: 3` field, along with `SenderMessagingPublicKey`, `SenderMessagingKeyName`, `RecipientMessagingPublicKey`, and `RecipientMessagingKeyName` fields for each message entry.
```
Comment (maebeam) [inline dips/dip-5.md:46]: Is it a good idea for `encrypt` to make a backend API request? To my knowledge none of the Identity APIs block on backend requests?

File: dips/dip-5.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+We also change how `PrivateMessage` transactions are connected. To send a V3 message, a `PrivateMessage` transaction should have `Version: 3` field in metadata, as well as `ExtraData` containing a message party information comprising `SenderMessagingPublicKey`, `SenderMessagingKeyName`, `RecipientMessagingPublicKey`, and `RecipientMessagingKeyName`. These fields will describe the messaging keys for the sender and recipient used when encrypting the private message. The fields will be added by core in `_connectPrivateMessage()` which will use them to construct a message and message party entries. Removing the entries from `UtxoView` or DB will happen in `_disconnectPrivateMessage()`.
+
+#### Backend
+
+We modify both `/api/v0/get-messages-stateless` and `/api/v0/send-message-stateless` endpoints to include the messaging party information. We also add an endpoint to fetch user messaging keys.
+
+#### Identity
```
Comment (maebeam) [inline dips/dip-5.md:44]: I'd like to see some information about how the keys are derived deterministically.

**Comment by deso2021** [comment]
can this proposal support encrypted group chat not only P2P chat?

**Comment by AeonSw4n** [comment]
@deso2021 yes, great point! I'm currently working on updating the messaging protocol further to support the group chat functionality.

**Pull request merged**